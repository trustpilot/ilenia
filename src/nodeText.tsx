import * as React from 'react';
import { interpolate, Interpolations, Tag, useTranslations } from '.';

export interface Node {
  start?: string;
  end?: string;
  key?: string;
  render: (match: React.ReactNode, key: string) => React.ReactNode;
}

export interface NodeTextProps {
  id: string;
  interpolations?: Interpolations;
  tag?: Tag;
  nodes: Node[];
};

const escapeRegex = (str: string) => str.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1');

const nodeRegex = (startTag: string, endTag: string) =>
  new RegExp(`${escapeRegex(startTag)}(.*)${escapeRegex(endTag)}`, 'g');

const addDefaultValues = (node: Node) => ({
  ...node,
  start: node.start || '[NODE-BEGIN]',
  end: node.end || '[NODE-END]',
});

export const NodeText = ({ id, nodes, interpolations, tag = { start: '[', end: ']' } }: NodeTextProps) => {
  const [translations] = useTranslations();
  let inputString = translations[id];
  if (!inputString) {
    console.error(`Couldn't find '${id}' in the translations table`); // eslint-disable-line no-console
    return null;
  }
  const nodeInterpolations: { [keyof: string]: React.ReactNode } = {};

  nodes.map(addDefaultValues).forEach(({ start, end, render, key: customKey }, index) => {
    const regexp = nodeRegex(start, end);
    const key = `text-interpolation-${index}`;
    const matches = regexp.exec(inputString);
    if (matches === null) return;
    const [withTags, nodeChildren] = matches;

    nodeInterpolations[key] = render(
      interpolations ? interpolate(nodeChildren, interpolations) : nodeChildren,
      customKey ?? key
    );
    inputString = inputString.replace(withTags, `${tag.start}${key}${tag.end}`);
  });

  return <React.Fragment>{interpolate(inputString, { ...nodeInterpolations, ...interpolations }, tag)}</React.Fragment>;
};

// Example Use, uncomment to check //
// eslint-disable-next-line @typescript-eslint/no-unused-vars
// const Test: React.FC = () => {
//   return (
//     <NodeText
//       id="whatever"
//       nodes={[
//         {
//           start: '[LINK-BEGIN]',
//           end: '[LINK-END]',
//           render: translation => `my own ${translation}`,
//         },
//       ]}
//     />
//   );
// };
