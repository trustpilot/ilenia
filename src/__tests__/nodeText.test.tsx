import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { LocalizationProvider, NodeText, LocaleDate } from '..';
import { Node } from '../nodeText';

const translations = {
  test1: 'No nodes in the string',
  test2: 'A node [NODE-BEGIN]in the[NODE-END] middle',
  test3: 'A node [NODE-BEGIN]in the[NODE-END] middle and one [NODE2-BEGIN]in the end[NODE2-END]',
  test4: 'A node [NODE-BEGIN]in the[NODE-END] middle',
  test5: 'A node [NODE-BEGIN]in the[NODE-END] middle and a date: [fromDate]',
  test6: 'From [NODE-BEGIN][fromDate][NODE-END] to [toDate]',
};

test('Renders a string with no links in it', () => {
  const { asFragment } = render(
    <LocalizationProvider locale="en-US" translations={translations}>
      <NodeText id="test1" nodes={[]} />
    </LocalizationProvider>
  );

  expect(asFragment()).toMatchSnapshot();
});

test('Renders a string with one node in the middle', () => {
  const { asFragment } = render(
    <LocalizationProvider locale="en-US" translations={translations}>
      <NodeText id="test2" nodes={[{ render: (m, key) => <p key={key}>{m}</p> }]} />
    </LocalizationProvider>
  );

  expect(asFragment()).toMatchSnapshot();
});


test('Renders a string with one node and custom tags', () => {
  const { asFragment } = render(
    <LocalizationProvider locale="en-US" translations={translations}>
      <NodeText id="test2" nodes={[{ render: (m, key) => <p key={key}>{m}</p> }]} tag={{ start: '{', end: '}' }}/>
    </LocalizationProvider>
  );

  expect(asFragment()).toMatchSnapshot();
});

test('Renders a string with multiple nodes in it', () => {
  const nodes: Node[] = [
    {
      render: (m, key) => <p key={key}>{m}</p>,
    },
    {
      start: '[NODE2-BEGIN]',
      end: '[NODE2-END]',
      render: (m, key) => <span key={key}>{m}</span>,
    },
  ];

  const { asFragment } = render(
    <LocalizationProvider locale="en-US" translations={translations}>
      <NodeText id="test3" nodes={nodes} />
    </LocalizationProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});

test('Renders a string with a working onClick callback', () => {
  const onClickSpy = jest.fn();

  const nodes: Node[] = [
    {
      render: (m, k) => <button onClick={onClickSpy} key={k}>{m}</button>
    },
  ];

  const { getByText, asFragment } = render(
    <LocalizationProvider locale="en-US" translations={translations}>
      <NodeText id="test4" nodes={nodes} />
    </LocalizationProvider>
  );

  fireEvent.click(getByText(/in the/i));

  expect(onClickSpy).toHaveBeenCalledTimes(1);
  expect(asFragment()).toMatchSnapshot();
});

test('Renders nothing when the string is not there', () => {
  console.error = jest.fn(); // eslint-disable-line no-console
  const { asFragment } = render(
    <LocalizationProvider locale="en-US" translations={translations}>
      <NodeText id="not-there" nodes={[]} />
    </LocalizationProvider>
  );

  expect(asFragment()).toMatchSnapshot();
  expect(console.error).toHaveBeenCalled(); // eslint-disable-line no-console
});

test('Renders a string with a node and an interpolated value', () => {
  const nodes: Node[] = [
    {
      render: (m, k) => <p key={k}>{m}</p>,
    },
  ];

  const interpolations = {
    fromDate: <LocaleDate date={new Date(2019, 1, 1)} />,
  };

  const { asFragment } = render(
    <LocalizationProvider locale="en-US" translations={translations}>
      <NodeText id="test5" nodes={nodes} interpolations={interpolations} />
    </LocalizationProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});

test('Renders a string with a node and an interpolated value inside, plus an interpolated value', () => {
  const nodes: Node[] = [
    {
      render: (m, k) => <p key={k}>{m}</p>,
    },
  ];

  const interpolations = {
    fromDate: <LocaleDate key="from-date" date={new Date(2019, 1, 1)} />,
    toDate: <LocaleDate key="to-date" date={new Date(2019, 1, 28)} />,
  };

  const { asFragment } = render(
    <LocalizationProvider locale="en-US" translations={translations}>
      <NodeText id="test6" nodes={nodes} interpolations={interpolations} />
    </LocalizationProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});
