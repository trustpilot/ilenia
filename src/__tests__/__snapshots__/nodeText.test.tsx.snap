// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders a string with a node and an interpolated value 1`] = `
<DocumentFragment>
  A node 
  <p>
    in the
  </p>
   middle and a date: 2/1/2019
</DocumentFragment>
`;

exports[`Renders a string with a node and an interpolated value inside, plus an interpolated value 1`] = `
<DocumentFragment>
  From 
  <p>
    2/1/2019
  </p>
   to 2/28/2019
</DocumentFragment>
`;

exports[`Renders a string with a working onClick callback 1`] = `
<DocumentFragment>
  A node 
  <button>
    in the
  </button>
   middle
</DocumentFragment>
`;

exports[`Renders a string with multiple nodes in it 1`] = `
<DocumentFragment>
  A node 
  <p>
    in the
  </p>
   middle and one 
  <span>
    in the end
  </span>
</DocumentFragment>
`;

exports[`Renders a string with no links in it 1`] = `
<DocumentFragment>
  No nodes in the string
</DocumentFragment>
`;

exports[`Renders a string with one node and custom tags 1`] = `
<DocumentFragment>
  A node 
  <p>
    in the
  </p>
   middle
</DocumentFragment>
`;

exports[`Renders a string with one node in the middle 1`] = `
<DocumentFragment>
  A node 
  <p>
    in the
  </p>
   middle
</DocumentFragment>
`;

exports[`Renders nothing when the string is not there 1`] = `<DocumentFragment />`;
